util/EIP-152/src/avx2.rs

```diff
@@ -0,0 +1,417 @@
+}
+
+#[inline(always)]
+unsafe fn xor(a: __m256i, b: __m256i) -> __m256i {
+	_mm256_xor_si256(a, b)
+}
+
+#[inline(always)]
+unsafe fn set4(a: u64, b: u64, c: u64, d: u64) -> __m256i {
+	_mm256_setr_epi64x(a as i64, b as i64, c as i64, d as i64)
+}
+
+// Adapted from https://github.com/rust-lang-nursery/stdsimd/pull/479.
+macro_rules! _MM_SHUFFLE {
+    ($z:expr, $y:expr, $x:expr, $w:expr) => {
+        ($z << 6) | ($y << 4) | ($x << 2) | $w
+    };
+}
+
+#[inline(always)]
+unsafe fn rotate_right_32(x: __m256i) -> __m256i {
+	_mm256_shuffle_epi32(x, _MM_SHUFFLE!(2, 3, 0, 1))
+}
+
+#[inline(always)]
+unsafe fn rotate_right_24(x: __m256i) -> __m256i {
+	let rotate24 = _mm256_setr_epi8(
+		3, 4, 5, 6, 7, 0, 1, 2, 11, 12, 13, 14, 15, 8, 9, 10, 3, 4, 5, 6, 7, 0, 1, 2, 11, 12, 13,
+		14, 15, 8, 9, 10,
+	);
+	_mm256_shuffle_epi8(x, rotate24)
+}
+
+#[inline(always)]
+unsafe fn rotate_right_16(x: __m256i) -> __m256i {
+	let rotate16 = _mm256_setr_epi8(
+		2, 3, 4, 5, 6, 7, 0, 1, 10, 11, 12, 13, 14, 15, 8, 9, 2, 3, 4, 5, 6, 7, 0, 1, 10, 11, 12,
+		13, 14, 15, 8, 9,
+	);
+	_mm256_shuffle_epi8(x, rotate16)
+}
+
+#[inline(always)]
+unsafe fn rotate_right_63(x: __m256i) -> __m256i {
+	_mm256_or_si256(_mm256_srli_epi64(x, 63), add(x, x))
+}
+
+#[inline(always)]
+unsafe fn g1(a: &mut __m256i, b: &mut __m256i, c: &mut __m256i, d: &mut __m256i, m: &mut __m256i) {
+	*a = add(*a, *m);
+	*a = add(*a, *b);
+	*d = xor(*d, *a);
+	*d = rotate_right_32(*d);
+	*c = add(*c, *d);
+	*b = xor(*b, *c);
+	*b = rotate_right_24(*b);
+}
+
+#[inline(always)]
+unsafe fn g2(a: &mut __m256i, b: &mut __m256i, c: &mut __m256i, d: &mut __m256i, m: &mut __m256i) {
+	*a = add(*a, *m);
+	*a = add(*a, *b);
+	*d = xor(*d, *a);
+	*d = rotate_right_16(*d);
+	*c = add(*c, *d);
+	*b = xor(*b, *c);
+	*b = rotate_right_63(*b);
+}
+
+// Note the optimization here of leaving b as the unrotated row, rather than a.
+// All the message loads below are adjusted to compensate for this. See
+// discussion at https://github.com/sneves/blake2-avx2/pull/4
+#[inline(always)]
+unsafe fn diagonalize(a: &mut __m256i, _b: &mut __m256i, c: &mut __m256i, d: &mut __m256i) {
+	*a = _mm256_permute4x64_epi64(*a, _MM_SHUFFLE!(2, 1, 0, 3));
+	*d = _mm256_permute4x64_epi64(*d, _MM_SHUFFLE!(1, 0, 3, 2));
+	*c = _mm256_permute4x64_epi64(*c, _MM_SHUFFLE!(0, 3, 2, 1));
+}
+
+#[inline(always)]
+unsafe fn undiagonalize(a: &mut __m256i, _b: &mut __m256i, c: &mut __m256i, d: &mut __m256i) {
+	*a = _mm256_permute4x64_epi64(*a, _MM_SHUFFLE!(0, 3, 2, 1));
+	*d = _mm256_permute4x64_epi64(*d, _MM_SHUFFLE!(1, 0, 3, 2));
+	*c = _mm256_permute4x64_epi64(*c, _MM_SHUFFLE!(2, 1, 0, 3));
+}
+
+
+/// The Blake2b compression function F. See https://tools.ietf.org/html/rfc7693#section-3.2
+/// Takes as an argument the state vector `state`, message block vector `message`, offset counter, final
+/// block indicator flag `f`, and number of rounds `rounds`. The state vector provided as the first
+/// parameter is modified by the function.
```

Comment:
It would be good to include an overview of how the AVX2 impl works here too. It looks very different from the pseudo-code in the spec so I think it's useful to have to give future readers a chance to follow the code.